back-end/java/main/java/src/Filtering/Node_KDTree/KDTree.java

```diff
@@ -0,0 +1,227 @@
+package src.Filtering.Node_KDTree;
+
+import java.util.Set;
+import src.DormRoom.DormBuilding;
+import src.DormRoom.BathroomType;
+import src.DormRoom.DormRoom;
+import src.Filtering.FilteringCriteria;
+import src.Filtering.IDormFilter;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+/**
+ * The {@code KDTree} class is a hierarchical data structure that organizes
+ * {@link DormRoom} instances according to multiple categorical attributes:
+ * {@link DormBuilding}, suite availability, kitchen availability, and
+ * {@link BathroomType} respectively.
+ *
+ * <p>Internally, it maintains a nested {@link HashMap} structure keyed by these
+ * categorical attributes, ultimately storing rooms in {@link KDTreeNode} instances, which
+ * themselves manage spatial (k=3 dimensions) searching of rooms based on the
+ * square footage, people per washer, and room capacity.
+ *
+ * <p>The outer layers of these nested maps categorize rooms, while the
+ * {@link KDTreeNode} at the leaves is responsible for efficient quantitative
+ * queries and filtering.
+ *
+ * <p>This class implements the {@link IDormFilter} interface, allowing it to
+ * be integrated with other dorm filtering systems.
+ */
+public class KDTree implements IDormFilter {
```

Comment:
So, a lot to talk about here:

0) One small thing, this is not a KDTree. KDTreeNode is a KDTree while this is a Tree where the "leaves" are KDTrees. I would change the class name to reflect this.

1) Firstly I want to say that it looks like everything works. For the repeated nesting for filter and building the hashmap I stopped reading after like 1-2 nests as it's pretty much the same the whole way down and if you did the same thing it should work.

2) Super nested hash maps are effectively the same thing as a tree. However, it is often better that if you are using beyond 1-2 nests then you should use a tree, as it is almost always much easier to read. We have 3-4 nests here, so the code becomes very annoying to read and at the very least the steps should be broken down by separate methods.  

3) The reason KDTree is not recommended for categorical variables is because for nearest neighbor searching it does not make sense, as categorical variables have constant distances. Nearest neighbor searching is by far the most common reason to use KDTrees so that is why it is not recommended to use categorical variables for them. For range searches, using categorical variables is fine (although a little slower, but with a very small number of categorical examples we should be fine. Also "slower" is still very fast regardless of size). However, our program does not implement nearest neighbor searching, so KDTree should work with all variables.

3.1) This structure is significantly more complicated due to the nested hash maps with KDTrees. Yes, it's faster, but the bulk of the runtime still comes from the KDTree range searches on numerical variables. The code would be much easier to read if every variables was in KDTreeNode. This would also make it easier to sell to Brown, since it seems we eventually will actually be talking to engineers. 

 3.2) This code makes it difficult to modify, which is also very important for this project for future maintainability and just general good coding practice. Adding a new categorical variable is much harder than it would be if everything were just in KDTreeNode. 

4) We may need to change isSuite to an enum (but probably not), and add campus location to filter on. 

5) I feel like I'm forgetting some points so I'm just noting that here